src/libusermode/running.cpp

```diff
@@ -0,0 +1,499 @@
+ * reuse, modify, and relicense the code.  DRAKVUF will always be          *
+ * available Open Source, but this is important because the inability to   *
+ * relicense code has caused devastating problems for other Free Software  *
+ * projects (such as KDE and NASM).                                        *
+ * To specify special license conditions of your contributions, just say   *
+ * so when you send them.                                                  *
+ *                                                                         *
+ * This program is distributed in the hope that it will be useful, but     *
+ * WITHOUT ANY WARRANTY; without even the implied warranty of              *
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the DRAKVUF   *
+ * license file for more details (it's in a COPYING file included with     *
+ * DRAKVUF, and also available from                                        *
+ * https://github.com/tklengyel/drakvuf/COPYING)                           *
+ *                                                                         *
+ ***************************************************************************/
+
+
+#include <fstream>
+#include <sstream>
+#include <map>
+#include <string>
+#include <optional>
+
+#include <config.h>
+#include <glib.h>
+#include <inttypes.h>
+#include <libvmi/libvmi.h>
+#include <libvmi/peparse.h>
+#include <libdrakvuf/libdrakvuf.h>
+#include <assert.h>
+
+#include "userhook.hpp"
+#include "uh-private.hpp"
+
+
+/**
+ * Running hook helper data structure.
+ * Used for passing user arguments and additional data between callbacks.
+ */
+struct rh_data_t
+{
+    // Arguments provided by the user.
+    addr_t target_process;
+    std::string dll_name;
+    std::string func_name;
+    callback_t cb;
+    void* extra;
+
+    // Additional data. Stored here for optimalization.
+    target_hook_state state;
+    vmi_pid_t target_process_pid;
+    addr_t target_process_dtb;
+    addr_t func_addr;
+
+    // We need to pass this around as we need offsets.
+    userhook* userhook_plugin;
+
+    rh_data_t(userhook* userhook_plugin, addr_t target_process, vmi_pid_t target_process_pid,
+              std::string dll_name, std::string func_name, callback_t cb, void* extra):
+        target_process(target_process), dll_name(dll_name), func_name(func_name), cb(cb),
+        extra(extra), state(HOOK_FIRST_TRY), target_process_pid(target_process_pid),
+        userhook_plugin(userhook_plugin) {}
+
+
+    static void free_trap(drakvuf_trap_t* trap)
+    {
+        if (!trap)
+            return;
+
+        if (trap->data)
+            delete (rh_data_t*) trap->data;
+
+        delete trap;
+    }
+};
+
+
+/**
+ * Searches process's InLoadOrderModuleList for given dll library and
+ * sets res_dll_base to its base address. Returns true on success.
+ */
+static
+bool get_dll_base(
+    drakvuf_t drakvuf,
+    const size_t* offsets,
```

Comment:
this is still an array decaying to pointer...
```suggestion
    const std::array<size_t, __OFFSET_MAX>& offsets,
```